pkg/controller/oam/applicationconfiguration/apply.go

```diff
@@ -19,73 +19,80 @@ package applicationconfiguration
 import (
 	"context"
 
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-	"k8s.io/apimachinery/pkg/types"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-
 	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
 	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
 	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/crossplane/oam-kubernetes-runtime/apis/core/v1alpha2"
-
-	"github.com/pkg/errors"
+	"github.com/crossplane/oam-kubernetes-runtime/pkg/oam/util"
 )
 
 // Reconcile error strings.
 const (
-	errFmtApplyWorkload  = "cannot apply workload %q"
-	errFmtSetWorkloadRef = "cannot set trait %q reference to %q"
-	errFmtApplyTrait     = "cannot apply trait %q %q"
-	errFmtApplyScope     = "cannot apply scope %q"
+	errFmtApplyWorkload      = "cannot apply workload %q"
+	errFmtSetWorkloadRef     = "cannot set trait %q reference to %q"
+	errFmtGetTraitDefinition = "cannot find trait definition %q %q %q"
+	errFmtApplyTrait         = "cannot apply trait %q %q %q"
+	errFmtApplyScope         = "cannot apply scope %q %q %q"
 )
 
 // A WorkloadApplicator creates or updates workloads and their traits.
 type WorkloadApplicator interface {
 	// Apply a workload and its traits.
-	Apply(ctx context.Context, namespace string, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error
+	Apply(ctx context.Context, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error
 }
 
 // A WorkloadApplyFn creates or updates workloads and their traits.
-type WorkloadApplyFn func(ctx context.Context, namespace string, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error
+type WorkloadApplyFn func(ctx context.Context, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error
 
 // Apply a workload and its traits.
-func (fn WorkloadApplyFn) Apply(ctx context.Context, namespace string, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error {
-	return fn(ctx, namespace, status, w, ao...)
+func (fn WorkloadApplyFn) Apply(ctx context.Context, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error {
+	return fn(ctx, status, w, ao...)
 }
 
 type workloads struct {
 	client    resource.Applicator
 	rawClient client.Client
 }
 
-func (a *workloads) Apply(ctx context.Context, namespace string, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error {
+func (a *workloads) Apply(ctx context.Context, status []v1alpha2.WorkloadStatus, w []Workload, ao ...resource.ApplyOption) error {
+	// they are all in the same namespace
+	var namespace = w[0].Workload.GetNamespace()
```

Comment:
what  if workload slice is empty?

Reply:
Is that even legal? I am assuming that we will have an admission hook